README.md
# IsoLang

`IsoLang` is a package for converting between ISO-639-1, ISO-639-2T, ISO-639-2B codes and the human readable name.

## Usage

```elixir
iex> IsoLang.all()
[
  %IsoLang{
    alpha2: "aa",
    alpha3b: "aar",
    alpha3t: "",
    name: "Afar",
    native_name: "Afaraf"
  },
  %IsoLang{
    alpha2: "ab",
    alpha3b: "abk",
    alpha3t: "",
    name: "Abkhazian",
    native_name: "аҧсуа бызшәа"
  },
  # ... etc...
]

# To get a single language, use `IsoLang.get/2` or its `!` counterpart.
# This will seek a match by examining the attributes `:alpha2`, `:alpha3b`, 
# `:alpha3t`, `:name` in that order:
iex> IsoLang.get("de")
{:ok,
 %IsoLang{
   alpha2: "de",
   alpha3b: "ger",
   alpha3t: "deu",
   name: "German",
   native_name: "Deutsch"
 }}

# Or to search just a single field, include the `:by` argument (capitalization matters):
iex(1)> IsoLang.get("French")
{:ok,%IsoLang{
   alpha2: "fr",
   alpha3b: "fre",
   alpha3t: "fra",
   name: "French",
   native_name: "Français"
 }}
```

Or search for matching languages using `IsoLang.find/2`:

```elixir
iex> IsoLang.find("eng")
{:ok,
  [
    %IsoLang{alpha2: "bn", alpha3b: "ben", alpha3t: "", name: "Bengali"},
    %IsoLang{alpha2: "en", alpha3b: "eng", alpha3t: "", name: "English"}
  ]}
```

The `native_name`s of the languages are translated via `gettext`: if you you need to display the names of languages in a specific language, you may add a dedicated `.po` file.

## Installation

```elixir
def deps do
  [
    {:iso_lang, "~> 0.3.1"}
  ]
end
```

## See Also

- <https://en.wikipedia.org/wiki/IETF_language_tag>
- <https://tools.ietf.org/search/bcp47>
- <https://datahub.io/core/language-codes#resource-language-codes-full>


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.4.0

- Moves `use Gettext` inside "private" module so we don't pollute the `IsoLang` namespace with all the generated functions.
- Sets `IsoLang.Data` to `@moduledoc false` to better communicate that it is an implementation detail without stand-alone functionality

## v0.3.1

- Adds `CHANGELOG`
- Defines `@external_resource` for `IsoLang.Data` to link CSV asset
- Updates examples in `@doc` blocks
- Updates dependencies to latest

## previous versions

The beforetimes...


test/iso_lang_test.exs
defmodule IsoLangTest do
  use ExUnit.Case

  describe "get/2" do
    test "Gets language when no standard is set" do
      assert {:ok, %IsoLang{alpha2: "de", alpha3b: "ger", alpha3t: "deu", name: "German"}} =
               IsoLang.get("de")
    end

    test "Gets language by alpha3b standard" do
      assert {:ok, %IsoLang{alpha2: "ko", alpha3b: "kor", alpha3t: "", name: "Korean"}} =
               IsoLang.get("kor", by: :alpha3b)
    end

    test "Gets language by alpha3t standard" do
      assert {:ok, %IsoLang{alpha2: "de", alpha3b: "ger", alpha3t: "deu", name: "German"}} =
               IsoLang.get("deu", by: :alpha3t)
    end

    test "Language automatically matches alpha3t" do
      assert {:ok, %IsoLang{alpha2: "de", alpha3b: "ger", alpha3t: "deu", name: "German"}} =
               IsoLang.get("deu")
    end

    test "Error when no language is found" do
      assert {:error, _} = IsoLang.get("invalid")
    end
  end

  describe "get!/2" do
    test "gets language when exists" do
      assert %IsoLang{alpha2: "de", alpha3b: "ger", alpha3t: "deu", name: "German"} =
               IsoLang.get!("de")
    end

    test "raises when no language is found" do
      assert_raise RuntimeError, fn -> IsoLang.get!("invalid") end
    end
  end

  describe "find/2" do
    test "Match partial names" do
      assert {:ok,
              [
                %IsoLang{
                  alpha2: "",
                  alpha3b: "ang",
                  alpha3t: "",
                  name: "English, Old (ca.450-1100)"
                },
                %IsoLang{alpha2: "bn", alpha3b: "ben", alpha3t: "", name: "Bengali"},
                %IsoLang{
                  alpha2: "",
                  alpha3b: "cpe",
                  alpha3t: "",
                  name: "Creoles and pidgins, English based"
                },
                %IsoLang{alpha2: "en", alpha3b: "eng", alpha3t: "", name: "English"},
                %IsoLang{
                  alpha2: "",
                  alpha3b: "enm",
                  alpha3t: "",
                  name: "English, Middle (1100-1500)"
                }
              ]} = IsoLang.find("eng")
    end

    test "Matches based on standard" do
      assert {:ok, [%IsoLang{alpha2: "ko", alpha3b: "kor", alpha3t: "", name: "Korean"}]} =
               IsoLang.find("ko", by: :alpha2)
    end

    test "Empty list when no match is found in standard" do
      assert {:ok, []} = IsoLang.find("eng", by: :alpha2)
    end
  end

  describe "find!/2" do
    test "match on partial names" do
      assert [
               %IsoLang{
                 alpha2: "",
                 alpha3b: "ang",
                 alpha3t: "",
                 name: "English, Old (ca.450-1100)"
               },
               %IsoLang{alpha2: "bn", alpha3b: "ben", alpha3t: "", name: "Bengali"},
               %IsoLang{
                 alpha2: "",
                 alpha3b: "cpe",
                 alpha3t: "",
                 name: "Creoles and pidgins, English based"
               },
               %IsoLang{alpha2: "en", alpha3b: "eng", alpha3t: "", name: "English"},
               %IsoLang{
                 alpha2: "",
                 alpha3b: "enm",
                 alpha3t: "",
                 name: "English, Middle (1100-1500)"
               }
             ] = IsoLang.find!("eng")
    end
  end
end


mix.exs
defmodule IsoLang.MixProject do
  use Mix.Project

  @source_url "https://github.com/luoken/iso_lang"
  @version "0.4.0"

  def project do
    [
      app: :iso_lang,
      version: @version,
      elixir: "~> 1.13",
      start_permanent: Mix.env() == :prod,
      aliases: aliases(),
      deps: deps(),
      description:
        "Formalizes interface for ISO-639 language codes and assists with language-code conversion",
      package: package(),
      elixirc_paths: elixirc_paths(Mix.env()),
      elixirc_options: [
        warnings_as_errors: true
      ],
      test_coverage: [
        tool: ExCoveralls
      ],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        credo: :test,
        lint: :test,
        "local.hex": :prod,
        "local.rebar": :prod,
        "deps.get": :prod,
        test_reset: :test
      ],
      docs: [
        main: "readme",
        source_ref: "v#{@version}",
        source_url: @source_url,
        extras: extras(),
        skip_undefined_reference_warnings_on: ["CHANGELOG.md"]
      ]
    ]
  end

  defp aliases do
    [
      lint: ["format --check-formatted", "credo --strict"]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:credo, "~> 1.7", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.3", only: [:dev], runtime: false},
      {:ex_doc, "~> 0.29", only: [:dev, :test], runtime: false},
      {:excoveralls, "~> 0.16", only: [:dev, :test], runtime: false},
      {:gettext, ">= 0.19.1"}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  defp package do
    [
      maintainers: ["Everett Griffiths", "Ken Luo"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/luoken/iso_lang"}
    ]
  end

  def extras do
    [
      "README.md",
      "cheatsheet.cheatmd",
      "CHANGELOG.md"
    ]
  end
end


lib/iso_lang/data.ex
defmodule IsoLang.Data do
  @moduledoc false
  # Builds list of language codes at compile-time.
  #
  # Data sourced from https://datahub.io/core/language-codes#resource-language-codes-full
  # Native language names adapted from https://github.com/meikidd/iso-639-1/blob/master/src/data.js
  #
  # External resource used by this module
  # https://hexdocs.pm/elixir/Module.html#module-external_resource
  @external_resource Application.app_dir(:iso_lang, ["priv", "language-codes-full.csv"])

  native_names = %{
    "aa" => "Afaraf",
    "ab" => "аҧсуа бызшәа",
    "ae" => "avesta",
    "af" => "Afrikaans",
    "ak" => "Akan",
    "am" => "አማርኛ",
    "an" => "aragonés",
    "ar" => "اَلْعَرَبِيَّةُ",
    "as" => "অসমীয়া",
    "av" => "авар мацӀ",
    "ay" => "aymar aru",
    "az" => "azərbaycan dili",
    "ba" => "башҡорт теле",
    "be" => "беларуская мова",
    "bg" => "български език",
    "bi" => "Bislama",
    "bm" => "bamanankan",
    "bn" => "বাংলা",
    "bo" => "བོད་ཡིག",
    "br" => "brezhoneg",
    "bs" => "bosanski jezik",
    "ca" => "Català",
    "ce" => "нохчийн мотт",
    "ch" => "Chamoru",
    "co" => "corsu",
    "cr" => "ᓀᐦᐃᔭᐍᐏᐣ",
    "cs" => "čeština",
    "cu" => "ѩзыкъ словѣньскъ",
    "cv" => "чӑваш чӗлхи",
    "cy" => "Cymraeg",
    "da" => "dansk",
    "de" => "Deutsch",
    "dv" => "ދިވެހި",
    "dz" => "རྫོང་ཁ",
    "ee" => "Eʋegbe",
    "el" => "Ελληνικά",
    "en" => "English",
    "eo" => "Esperanto",
    "es" => "Español",
    "et" => "eesti",
    "eu" => "euskara",
    "fa" => "فارسی",
    "ff" => "Fulfulde",
    "fi" => "suomi",
    "fj" => "vosa Vakaviti",
    "fo" => "føroyskt",
    "fr" => "Français",
    "fy" => "Frysk",
    "ga" => "Gaeilge",
    "gd" => "Gàidhlig",
    "gl" => "galego",
    "gn" => "Avañe'ẽ",
    "gu" => "ગુજરાતી",
    "gv" => "Gaelg",
    "ha" => "هَوُسَ",
    "he" => "עברית",
    "hi" => "हिन्दी",
    "ho" => "Hiri Motu",
    "hr" => "Hrvatski",
    "ht" => "Kreyòl ayisyen",
    "hu" => "magyar",
    "hy" => "Հայերեն",
    "hz" => "Otjiherero",
    "ia" => "Interlingua",
    "id" => "Bahasa Indonesia",
    "ie" => "Interlingue",
    "ig" => "Asụsụ Igbo",
    "ii" => "ꆈꌠ꒿ Nuosuhxop",
    "ik" => "Iñupiaq",
    "io" => "Ido",
    "is" => "Íslenska",
    "it" => "Italiano",
    "iu" => "ᐃᓄᒃᑎᑐᑦ",
    "ja" => "日本語",
    "jv" => "basa Jawa",
    "ka" => "ქართული",
    "kg" => "Kikongo",
    "ki" => "Gĩkũyũ",
    "kj" => "Kuanyama",
    "kk" => "қазақ тілі",
    "kl" => "kalaallisut",
    "km" => "ខេមរភាសា",
    "kn" => "ಕನ್ನಡ",
    "ko" => "한국어",
    "kr" => "Kanuri",
    "ks" => "कश्मीरी",
    "ku" => "Kurdî",
    "kv" => "коми кыв",
    "kw" => "Kernewek",
    "ky" => "Кыргызча",
    "la" => "latine",
    "lb" => "Lëtzebuergesch",
    "lg" => "Luganda",
    "li" => "Limburgs",
    "ln" => "Lingála",
    "lo" => "ພາສາລາວ",
    "lt" => "lietuvių kalba",
    "lu" => "Kiluba",
    "lv" => "latviešu valoda",
    "mg" => "fiteny malagasy",
    "mh" => "Kajin M̧ajeļ",
    "mi" => "te reo Māori",
    "mk" => "македонски јазик",
    "ml" => "മലയാളം",
    "mn" => "Монгол хэл",
    "mr" => "मराठी",
    "ms" => "Bahasa Melayu",
    "mt" => "Malti",
    "my" => "ဗမာစာ",
    "na" => "Dorerin Naoero",
    "nb" => "Norsk bokmål",
    "nd" => "isiNdebele",
    "ne" => "नेपाली",
    "ng" => "Owambo",
    "nl" => "Nederlands",
    "nn" => "Norsk nynorsk",
    "no" => "Norsk",
    "nr" => "isiNdebele",
    "nv" => "Diné bizaad",
    "ny" => "chiCheŵa",
    "oc" => "occitan",
    "oj" => "ᐊᓂᔑᓈᐯᒧᐎᓐ",
    "om" => "Afaan Oromoo",
    "or" => "ଓଡ଼ିଆ",
    "os" => "ирон æвзаг",
    "pa" => "ਪੰਜਾਬੀ",
    "pi" => "पाऴि",
    "pl" => "Polski",
    "ps" => "پښتو",
    "pt" => "Português",
    "qu" => "Runa Simi",
    "rm" => "rumantsch grischun",
    "rn" => "Ikirundi",
    "ro" => "Română",
    "ru" => "Русский",
    "rw" => "Ikinyarwanda",
    "sa" => "संस्कृतम्",
    "sc" => "sardu",
    "sd" => "सिन्धी",
    "se" => "Davvisámegiella",
    "sg" => "yângâ tî sängö",
    "si" => "සිංහල",
    "sk" => "slovenčina",
    "sl" => "slovenščina",
    "sm" => "gagana fa'a Samoa",
    "sn" => "chiShona",
    "so" => "Soomaaliga",
    "sq" => "Shqip",
    "sr" => "српски језик",
    "ss" => "SiSwati",
    "st" => "Sesotho",
    "su" => "Basa Sunda",
    "sv" => "Svenska",
    "sw" => "Kiswahili",
    "ta" => "தமிழ்",
    "te" => "తెలుగు",
    "tg" => "тоҷикӣ",
    "th" => "ไทย",
    "ti" => "ትግርኛ",
    "tk" => "Türkmençe",
    "tl" => "Wikang Tagalog",
    "tn" => "Setswana",
    "to" => "faka Tonga",
    "tr" => "Türkçe",
    "ts" => "Xitsonga",
    "tt" => "татар теле",
    "tw" => "Twi",
    "ty" => "Reo Tahiti",
    "ug" => "ئۇيغۇرچە‎",
    "uk" => "Українська",
    "ur" => "اردو",
    "uz" => "Ўзбек",
    "ve" => "Tshivenḓa",
    "vi" => "Tiếng Việt",
    "vo" => "Volapük",
    "wa" => "walon",
    "wo" => "Wollof",
    "xh" => "isiXhosa",
    "yi" => "ייִדיש",
    "yo" => "Yorùbá",
    "za" => "Saɯ cueŋƅ",
    "zh" => "中文",
    "zu" => "isiZulu"
  }

  lines =
    @external_resource
    |> File.stream!()
    |> Stream.drop(1)

  langs =
    Enum.map(lines, fn l ->
      [alpha3b, alpha3t, alpha2 | rest] =
        l
        |> String.trim()
        |> String.split(",")

      tail = Enum.join(rest, ",")

      # Parse up to the closing double-quote or up to the comma
      f = fn
        <<?", chunk::binary>> ->
          Regex.run(~r/(.*)"/U, chunk, capture: :all_but_first) |> Enum.at(0)

        chunk ->
          Regex.run(~r/(.*),/U, chunk, capture: :all_but_first) |> Enum.at(0)
      end

      %IsoLang{
        alpha2: alpha2,
        alpha3b: alpha3b,
        alpha3t: alpha3t,
        name: Gettext.gettext(IsoLang.Backend, f.(tail)),
        native_name: Map.get(native_names, alpha2)
      }
    end)

  def all(_opts) do
    unquote(Macro.escape(langs))
  end
end


lib/iso_lang.ex
defmodule IsoLang do
  @moduledoc """
  Documentation for `IsoLang`.

  Provides utilities for dealing with [ISO 639](https://en.wikipedia.org/wiki/ISO_639)
  languages.

  ## See Also

  - https://en.wikipedia.org/wiki/IETF_language_tag
  - https://tools.ietf.org/search/bcp47
  - https://datahub.io/core/language-codes#resource-language-codes-full
  """
  @type t :: %__MODULE__{
          alpha2: String.t(),
          alpha3b: String.t(),
          alpha3t: String.t(),
          name: String.t(),
          native_name: String.t()
        }

  defstruct alpha2: nil, alpha3b: nil, alpha3t: nil, name: nil, native_name: nil

  defmodule Backend do
    @moduledoc false
    # We rely on a dedicated "private" module as the Gettext backend so we don't
    # pollute the IsoLang docs with all the generated Gettext functions
    use Gettext, otp_app: :iso_lang
  end

  @doc """
  Returns a list of all available ISO language codes.
  """
  def all(opts \\ []), do: IsoLang.Data.all(opts)

  @doc """
  Gets a single language struct identified by a field.
  If the `:by` field is not specified, fields are checked in the following order:

  - `:alpha2`
  - `:alpha3b`
  - `:alpha3t`
  - `:name`

  Keep in mind that the `:alpha2` (2-character codes) and `:alpha3b` (3-character codes)
  are the most common -- not every language defines a `:alpha3t` code.

  ## Options

  - `:by` specifies which struct field to be used in the search. One of
    `:alpha2`, `:alpha3b`, `:alpha3t`, `:name`. (optional; default will examine all
    fields for a match in the order indicated above)

  ## Examples

      iex> IsoLang.get("de")
      {:ok,
        %IsoLang{
          alpha2: "de",
          alpha3b: "ger",
          alpha3t: "deu",
          name: "German",
          native_name: "Deutsch"
        }}
  """
  @spec get(value :: String.t(), opts :: Keyword.t()) :: {:ok, IsoLang.t()} | {:error, any()}
  def get(value, opts \\ []) do
    key = Keyword.get(opts, :by, nil)

    opts
    |> all()
    |> Enum.find(:not_found, fn
      lang when not is_nil(key) -> Map.get(lang, key) == value
      %__MODULE__{alpha2: ^value} -> true
      %__MODULE__{alpha3b: ^value} -> true
      %__MODULE__{alpha3t: ^value} -> true
      %__MODULE__{name: ^value} -> true
      _ -> false
    end)
    |> case do
      :not_found -> {:error, "Language not found"}
      lang -> {:ok, lang}
    end
  end

  @doc """
  As `get/2`, but raises on error
  """
  def get!(query, opts \\ []) do
    case get(query, opts) do
      {:ok, lang} -> lang
      {:error, error} -> raise error
    end
  end

  @doc """
  Searches for matching languages using a case-insensitive query string

  ## Options

  - `:by` specifies which struct field to be used in the search. One of
    `:alpha2`, `:alpha3b`, `:alpha3t`, `:name`. Default: `:name`

  ## Examples

      iex> IsoLang.find("eng")
      {:ok,
        [
          %IsoLang{
            alpha2: "",
            alpha3b: "ang",
            alpha3t: "",
            name: "English, Old (ca.450-1100)",
            native_name: nil
          },
          %IsoLang{
            alpha2: "bn",
            alpha3b: "ben",
            alpha3t: "",
            name: "Bengali",
            native_name: "বাংলা"
          },
          # ... etc...
        ]
      }
  """
  @spec find(query :: String.t(), opts :: Keyword.t()) :: {:ok, [IsoLang.t()]} | {:error, any()}
  def find(query, opts \\ []) do
    key = Keyword.get(opts, :by, :name)

    {:ok,
     opts
     |> all()
     |> Enum.filter(fn lang ->
       query
       |> Regex.compile!("i")
       |> Regex.match?(
         lang
         |> Map.fetch!(key)
       )
     end)}
  rescue
    e -> {:error, e}
  end

  @doc """
  As `find/2`, but raises on error
  """
  def find!(query, opts \\ []) do
    case find(query, opts) do
      {:ok, lang} -> lang
      {:error, error} -> raise error
    end
  end
end
